File: arc/go/stl/channel/state.go
// Copyright 2026 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

package channel

import (
	"slices"

	"github.com/synnaxlabs/x/telem"
	xunsafe "github.com/synnaxlabs/x/unsafe"
)

// Digest provides metadata about a channel for state initialization.
type Digest struct {
	DataType telem.DataType
	Key      uint32
	Index    uint32
}

// ProgramState manages channel I/O buffers and index mapping.
type ProgramState struct {
	reads           map[uint32]telem.MultiSeries
	writes          map[uint32]telem.Series
	activeWriteKeys []uint32
	indexes         map[uint32]uint32
}

// NewProgramState creates a new ProgramState from channel digests.
func NewProgramState(digests []Digest) *ProgramState {
	cs := &ProgramState{
		reads:   make(map[uint32]telem.MultiSeries),
		writes:  make(map[uint32]telem.Series),
		indexes: make(map[uint32]uint32),
	}
	for _, d := range digests {
		cs.indexes[d.Key] = d.Index
	}
	return cs
}

// Ingest adds external channel data to the read buffer.
func (cs *ProgramState) Ingest(fr telem.Frame[uint32]) {
	for rawI, key := range fr.RawKeys() {
		if fr.ShouldExcludeRaw(rawI) {
			continue
		}
		cs.reads[key] = cs.reads[key].Append(fr.RawSeriesAt(rawI))
	}
}

// Flush extracts buffered channel writes into a frame and clears the write
// buffer. Only channels written in the current cycle are flushed.
func (cs *ProgramState) Flush(
	fr telem.Frame[uint32],
) (telem.Frame[uint32], bool) {
	if len(cs.activeWriteKeys) == 0 {
		return fr, false
	}
	flushed := false
	for _, key := range cs.activeWriteKeys {
		data, ok := cs.writes[key]
		if !ok || len(data.Data) == 0 {
			continue
		}
		fr = fr.Append(key, data.DeepCopy())
		flushed = true
		data.Data = data.Data[:0]
		data.TimeRange = telem.TimeRangeZero
		data.Alignment = 0
		cs.writes[key] = data
	}
	cs.activeWriteKeys = cs.activeWriteKeys[:0]
	return fr, flushed
}

// clearReadsReallocThreshold is the backing array capacity above which
// ClearReads allocates a fresh slice instead of re-slicing in place. Below
// this threshold, slices.Delete zeroes old references (allowing GC) without
// allocating.
const clearReadsReallocThreshold = 64

// ClearReads clears accumulated channel read buffers while preserving the
// latest series for each channel.
func (cs *ProgramState) ClearReads() {
	for key, ser := range cs.reads {
		if len(ser.Series) <= 1 {
			continue
		}
		if cap(ser.Series) > clearReadsReallocThreshold {
			ser.Series = []telem.Series{ser.Series[len(ser.Series)-1]}
		} else {
			ser.Series = slices.Delete(
				ser.Series, 0, len(ser.Series)-1,
			)
		}
		cs.reads[key] = ser
	}
}

// ReadValue reads a single value from a channel (for WASM runtime bindings).
func (cs *ProgramState) ReadValue(key uint32) (telem.Series, bool) {
	ms, ok := cs.reads[key]
	if !ok || len(ms.Series) == 0 {
		return telem.Series{}, false
	}
	return ms.Series[len(ms.Series)-1], ok
}

// writeValue writes a single value to a channel (for WASM runtime bindings).
func (cs *ProgramState) writeValue(key uint32, value telem.Series) {
	cs.appendWriteSeries(key, value)
	cs.writeIndexedTimestamp(key)
}

func (cs *ProgramState) WriteChannelU8(key uint32, v uint8) {
	appendFixedWriteSample(cs, key, v)
	cs.writeIndexedTimestamp(key)
}
func (cs *ProgramState) WriteChannelU16(key uint32, v uint16) {
	appendFixedWriteSample(cs, key, v)
	cs.writeIndexedTimestamp(key)
}
func (cs *ProgramState) WriteChannelU32(key uint32, v uint32) {
	appendFixedWriteSample(cs, key, v)
	cs.writeIndexedTimestamp(key)
}
func (cs *ProgramState) WriteChannelU64(key uint32, v uint64) {
	appendFixedWriteSample(cs, key, v)
	cs.writeIndexedTimestamp(key)
}
func (cs *ProgramState) WriteChannelI8(key uint32, v int8) {
	appendFixedWriteSample(cs, key, v)
	cs.writeIndexedTimestamp(key)
}
func (cs *ProgramState) WriteChannelI16(key uint32, v int16) {
	appendFixedWriteSample(cs, key, v)
	cs.writeIndexedTimestamp(key)
}
func (cs *ProgramState) WriteChannelI32(key uint32, v int32) {
	appendFixedWriteSample(cs, key, v)
	cs.writeIndexedTimestamp(key)
}
func (cs *ProgramState) WriteChannelI64(key uint32, v int64) {
	appendFixedWriteSample(cs, key, v)
	cs.writeIndexedTimestamp(key)
}
func (cs *ProgramState) WriteChannelF32(key uint32, v float32) {
	appendFixedWriteSample(cs, key, v)
	cs.writeIndexedTimestamp(key)
}
func (cs *ProgramState) WriteChannelF64(key uint32, v float64) {
	appendFixedWriteSample(cs, key, v)
	cs.writeIndexedTimestamp(key)
}

func (cs *ProgramState) writeIndexedTimestamp(key uint32) {
	idx := cs.indexes[key]
	if idx != 0 {
		appendFixedWriteSample(cs, idx, telem.Now())
	}
}

func appendFixedWriteSample[T telem.FixedSample](cs *ProgramState, key uint32, value T) {
	dt := telem.InferDataType[T]()
	acc, exists := cs.writes[key]
	if !exists {
		acc = telem.Series{DataType: dt}
	}
	if len(acc.Data) == 0 {
		cs.activeWriteKeys = append(cs.activeWriteKeys, key)
	}
	if acc.DataType == telem.UnknownT {
		acc.DataType = dt
	}
	if acc.DataType != dt && len(acc.Data) > 0 {
		cs.writes[key] = telem.NewSeriesV(value)
		return
	}
	den := int(dt.Density())
	sampleStart := len(acc.Data)
	acc.Data = slices.Grow(acc.Data, den)
	acc.Data = acc.Data[:sampleStart+den]
	xunsafe.CastSlice[byte, T](acc.Data)[sampleStart/den] = value
	cs.writes[key] = acc
}

// readSeries reads buffered data and time series from a channel.
func (cs *ProgramState) readSeries(
	key uint32,
) (data telem.MultiSeries, time telem.MultiSeries, ok bool) {
	data, ok = cs.reads[key]
	if !ok {
		return telem.MultiSeries{}, telem.MultiSeries{}, false
	}
	indexKey := cs.indexes[key]
	if indexKey == 0 {
		return data, telem.MultiSeries{}, len(data.Series) > 0
	}
	time, ok = cs.reads[indexKey]
	if !ok {
		return telem.MultiSeries{}, telem.MultiSeries{}, false
	}
	return data, time, len(time.Series) > 0 && len(data.Series) > 0
}

func (cs *ProgramState) writeChannel(key uint32, data, time telem.Series) {
	cs.appendWriteSeries(key, data)
	idx := cs.indexes[key]
	if idx != 0 {
		cs.appendWriteSeries(idx, time)
	}
}

func (cs *ProgramState) appendWriteSeries(key uint32, source telem.Series) {
	acc, exists := cs.writes[key]
	if !exists {
		acc = telem.Series{DataType: source.DataType}
	}
	if len(acc.Data) == 0 {
		cs.activeWriteKeys = append(cs.activeWriteKeys, key)
	}
	if acc.DataType == telem.UnknownT {
		acc.DataType = source.DataType
	}
	if len(source.Data) == 0 {
		cs.writes[key] = acc
		return
	}
	if acc.DataType != source.DataType && len(acc.Data) > 0 {
		acc = source.DeepCopy()
		cs.writes[key] = acc
		return
	}
	if len(acc.Data) == 0 {
		acc.TimeRange = source.TimeRange
		acc.Alignment = source.Alignment
	} else {
		if source.TimeRange.Start < acc.TimeRange.Start {
			acc.TimeRange.Start = source.TimeRange.Start
		}
		if source.TimeRange.End > acc.TimeRange.End {
			acc.TimeRange.End = source.TimeRange.End
		}
	}
	acc.Data = append(acc.Data, source.Data...)
	cs.writes[key] = acc
}

File: arc/cpp/stl/channel/state.cpp
// Copyright 2026 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

#include <ranges>

#include "arc/cpp/stl/channel/state.h"

namespace arc::stl::channel {

State::State(const std::vector<Digest> &digests) {
    for (const auto &digest: digests)
        this->indexes[digest.key] = digest.index;
}

void State::ingest(const x::telem::Frame &frame) {
    for (size_t i = 0; i < frame.size(); i++)
        this->reads[frame.channels->at(i)].push_back(
            x::mem::local_shared(std::move(frame.series->at(i)))
        );
}

std::pair<x::telem::MultiSeries, bool> State::read_value(const types::ChannelKey key) {
    const auto it = this->reads.find(key);
    if (it == this->reads.end() || it->second.empty())
        return {x::telem::MultiSeries{}, false};
    x::telem::MultiSeries ms;
    for (const auto &s: it->second)
        ms.series.push_back(s->deep_copy());
    return {std::move(ms), true};
}

static void append_to_write_buffer(Series &dest, const Series &src) {
    if (src == nullptr || src->empty()) return;
    if (dest == nullptr) {
        dest = x::mem::make_local_shared<x::telem::Series>(src->deep_copy());
        return;
    }
    if (dest->data_type() != src->data_type())
        throw std::runtime_error("cannot append series with mismatched data types");
    if (dest->data_type().is_variable()) {
        const auto old_time_range = dest->time_range;
        auto merged = dest->strings();
        auto incoming = src->strings();
        merged.insert(merged.end(), incoming.begin(), incoming.end());
        dest = x::mem::make_local_shared<x::telem::Series>(merged, dest->data_type());
        dest->time_range = old_time_range;
    } else {
        const auto required_cap = dest->size() + src->size();
        if (dest->cap() < required_cap) {
            const auto grown_cap = std::max(required_cap, dest->cap() * 2 + 1);
            auto grown = x::mem::make_local_shared<x::telem::Series>(
                dest->data_type(),
                grown_cap
            );
            grown->write(*dest);
            grown->time_range = dest->time_range;
            grown->alignment = dest->alignment;
            dest = std::move(grown);
        }
        dest->write(*src);
    }
    if (dest->time_range == x::telem::TimeRange()) {
        dest->time_range = src->time_range;
    } else {
        if (src->time_range.start < dest->time_range.start)
            dest->time_range.start = src->time_range.start;
        if (src->time_range.end > dest->time_range.end)
            dest->time_range.end = src->time_range.end;
    }
}

void State::write_value(
    const types::ChannelKey key,
    const Series &data,
    const Series &time
) {
    auto &data_buf = this->writes[key];
    if (data_buf == nullptr || data_buf->empty())
        this->active_write_keys.push_back(key);
    append_to_write_buffer(data_buf, data);
    if (const auto idx_iter = this->indexes.find(key);
        idx_iter != this->indexes.end() && idx_iter->second != 0) {
        auto &time_buf = this->writes[idx_iter->second];
        if (time_buf == nullptr || time_buf->empty())
            this->active_write_keys.push_back(idx_iter->second);
        append_to_write_buffer(time_buf, time);
    }
}

#define IMPL_WRITE_CHANNEL(suffix, cpptype, dt_const)                                  \
    void State::write_channel_##suffix(                                                \
        const types::ChannelKey key,                                                   \
        const cpptype value                                                            \
    ) {                                                                                \
        this->write_channel_typed(key, dt_const, value);                               \
    }

IMPL_WRITE_CHANNEL(u8, uint8_t, x::telem::UINT8_T)
IMPL_WRITE_CHANNEL(u16, uint16_t, x::telem::UINT16_T)
IMPL_WRITE_CHANNEL(u32, uint32_t, x::telem::UINT32_T)
IMPL_WRITE_CHANNEL(u64, uint64_t, x::telem::UINT64_T)
IMPL_WRITE_CHANNEL(i8, int8_t, x::telem::INT8_T)
IMPL_WRITE_CHANNEL(i16, int16_t, x::telem::INT16_T)
IMPL_WRITE_CHANNEL(i32, int32_t, x::telem::INT32_T)
IMPL_WRITE_CHANNEL(i64, int64_t, x::telem::INT64_T)
IMPL_WRITE_CHANNEL(f32, float, x::telem::FLOAT32_T)
IMPL_WRITE_CHANNEL(f64, double, x::telem::FLOAT64_T)

#undef IMPL_WRITE_CHANNEL

std::tuple<x::telem::MultiSeries, x::telem::MultiSeries, bool>
State::read_series(const types::ChannelKey key) {
    auto [data, ok] = this->read_value(key);
    if (!ok) return {x::telem::MultiSeries{}, x::telem::MultiSeries{}, false};
    const auto index_it = this->indexes.find(key);
    if (index_it == this->indexes.end() || index_it->second == 0) {
        const bool has_data = !data.series.empty();
        return {std::move(data), x::telem::MultiSeries{}, has_data};
    }
    auto [time, time_ok] = this->read_value(index_it->second);
    if (!time_ok) return {x::telem::MultiSeries{}, x::telem::MultiSeries{}, false};
    const bool has_data = !data.series.empty() && !time.series.empty();
    return {std::move(data), std::move(time), has_data};
}

void State::write_series(
    const types::ChannelKey key,
    const Series &data,
    const Series &time
) {
    this->write_value(key, data, time);
}

void State::flush_into(x::telem::Frame &out) {
    for (auto &series_vec: this->reads | std::views::values) {
        if (series_vec.size() <= 1) continue;
        auto last = std::move(series_vec.back());
        series_vec.clear();
        series_vec.push_back(std::move(last));
    }
    for (const auto key: this->active_write_keys) {
        auto it = this->writes.find(key);
        if (it == this->writes.end() || it->second == nullptr || it->second->empty())
            continue;
        out.emplace(key, it->second->shallow_copy());
        it->second->detach_buffer();
        it->second->time_range = x::telem::TimeRange();
        it->second->alignment = x::telem::Alignment();
    }
    this->active_write_keys.clear();
}

void State::reset() {
    this->reads.clear();
    this->writes.clear();
    this->active_write_keys.clear();
}

}
